### PR TITLE
Activate queue message behavior for more intuitive message order

### LINKED ISSUE #2702: Spike out more user-intuitive message order
### User Story

As a user, I want messages to show up in an intuitive order so I can have a familiar and pleasant user experience.

As a developer, I want to understand and communicate semantics of message ordering clock-value.

As a (dev)tester, I want to report bugs that are bugs and not expected behavior.

### Description

*Type*: Feature/Spike

*Summary*: Spike out mini buffer where messages aren't acked (received) and this is ideally visually represented.

#### Expected behavior
Re-order messages when they arrive in wrong order in short timespan and visually distinguish this.

#### Actual behavior
It can appear as if messages are "in wrong order" https://github.com/status-im/status-react/issues/2699 even though this is the correct semantics.

### Notes

From Slack:
TLDR if you receive message 3 first it is _received_ first. If you later, even 10ms later, get message 1 and 2 then you need to choose: do we update ordering based on information (change UI) or so we append? This is a design decision and we default to latter right now.

Event chain:
```
SEND 1
SEND 2
SEND 3
<indeterministic network>
RECV 3  - received first so committed first
RECV 2 - we know senders clock-value so we can choose to re-order based on this info (and careful reasoning so we preserve Lamport semantics, for example tempids/uncommitted messages)
RECV 1
```

Also investigate if this problem is reproducible, in which it case the behavior on _send_ side might be "wrong"/suboptimal. I.e. SEND 1 2 3 on client results in Whisper "publish" 3 2 1 due to faulty buffer or something.

Also think carefully about implications for group chat. Probably need short circuit when sending messages to preserve Lamport semantics.

### Resources
- Initial PR https://github.com/status-im/status-react/commit/db6b80d529f3b7512d4831b88246b60a86072b1f
- Relevant ns: https://github.com/status-im/status-react/blob/db6b80d529f3b7512d4831b88246b60a86072b1f/src/status_im/utils/clocks.cljs
- Relevant tests for invariants preserved: https://github.com/status-im/status-react/blob/develop/test/cljs/status_im/test/utils/clocks.cljs

### PR DESCRIPTION
**Pull request by oskarth**
Fixes #2702 

See https://github.com/status-im/status-react/pull/2743 for ground work.

### Summary

Intercept received messages and queue them

- Exposes behavior under QUEUE_MESSAGE_ENABLED flag that is turned off in
production.

- Delays messages to allow for reordering of latecoming messages up to 500ms

### Review notes (optional):
Normal review.

### Testing notes (optional):
1. Make a build with QUEUE_MESSAGE_ENABLED enabled and one without
2. Send lots of messages in 1on1 chat from one person, like `1 2 3 4 5 6 7 8 9`.
3a. With flag off you should eventually see ordering being different due to network uncertainty
3b. With flag on this should essentially be impossible to achieve (or way less likely)

status: ready

### CODE REVIEW

**Review by janherich — APPROVED** [review]

File: src/status_im/utils/pre_receiver.cljs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+(defn- earliest-clock-value-seen? [seen id clock-value]
+  (->> seen
+       (filter (fn [[_ x]] (= x id)))
+       sort
+       ffirst
+       (= clock-value)))
+
+;; TODO: Update docstring after add-message fixed
+(defn start!
```
Comment (yenda) [inline src/status_im/utils/pre_receiver.cljs]: Should we write a tech debt bounty extracting the rolling window async mechanism into async utils namespace ?

**Review by yenda — APPROVED** [review]

File: src/status_im/utils/pre_receiver.cljs [hunk 2]
```diff
@@ -0,0 +1,44 @@
+(ns status-im.utils.pre-receiver
+  (:require-macros [cljs.core.async.macros :as async])
+  (:require [cljs.core.async :as async]))
+
+;; See status-im.test.utils.pre-receiver for justification.
+
+;; TODO(oskarth): Doesn't belong here, mocking real add-msg - fix signature
+(defn- add-message [{:keys [id clock-value] :as msg}]
```
Comment (oskarth) [inline src/status_im/utils/pre_receiver.cljs]: Oops, remove this, not used

File: src/status_im/utils/pre_receiver.cljs (hunk 1 above)
Comment (oskarth) [reply src/status_im/utils/pre_receiver.cljs]: @yenda We could, I just don't see exactly how it'd be different than this namespace (without earliest-clock-value-seen? fn). What is the tech debt in your opinion?

Comment (oskarth) [reply src/status_im/utils/pre_receiver.cljs]: The delay-msg fn looks pretty canonical to me, and the pre-receiver could maybe be more generalized, but it also seems quite tight with specific problem. Do we have another use case than this in mind right now?

**Comment by oskarth** [comment]
Rebased on develop with other merged PR. One question is if we should put flag earlier in call chain, and then about possible generic fn that @yenda and @janherich brought up (separate issue). Moving to ready for testing.

**Comment by asemiankevich** [comment]
1. This is build with QUEUE_MESSAGE_ENABLED=1

Branch: status-im:feature/spike-message-order-2-#2702
Android: https://i.diawi.com/952PHz
iOS: https://i.diawi.com/8Upuv8

2. This is build with QUEUE_MESSAGE_ENABLED=0

Branch: status-im:feature/spike-message-order-2-#2702
Android: https://i.diawi.com/8zYYFe
iOS: https://i.diawi.com/jpES9n
